### PR TITLE
[Model] add flux2 klein

### PR DESCRIPTION
**Pull request by david6666666**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose
support [black-forest-labs/FLUX.2-klein](https://huggingface.co/black-forest-labs/FLUX.2-klein-9B)
## Test Plan
vLLM-Omni:
```
python examples/offline_inference/text_to_image/text_to_image.py \
  --model black-forest-labs/FLUX.2-klein-9B \
  --prompt "a photo of a forest with mist swirling around the tree trunks. The word 'FLUX.2' is painted over it in big, red brush strokes with visible texture" \
  --seed 42 \
  --cfg_scale 4.0 \
  --num_images_per_prompt 1 \
  --num_inference_steps 4 \
  --guidance_scale 1.0 \
  --height 768 \
  --width 1360 \
  --output outputs/output_9b_omni.png

```

online serving:
```
vllm serve black-forest-labs/FLUX.2-klein-9B --omni --port 8091

curl -s http://localhost:8091/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{
    "messages": [
      {"role": "user", "content": "A beautiful landscape painting"}
    ],
    "extra_body": {
      "height": 1024,
      "width": 1024,
      "num_inference_steps": 4,
      "true_cfg_scale": 4.0,
      "seed": 42
    }
  }' | jq -r '.choices[0].message.content[0].image_url.url' | cut -d',' -f2- | base64 -d > klein.png
```

## Test Result
vLLM-Omni:
output_9b_omni

```
python examples/offline_inference/image_to_image/image_edit.py \
  --model black-forest-labs/FLUX.2-klein-9B \
  --image output_9b_omni.png \
  --prompt "replace the trees in the image with cyberpunk buildings." \
  --output output_image_edit.png \
  --num_inference_steps 4 \
  --cfg_scale 4.0
```
output_image_edit

diffusers:
```
import torch
from diffusers import Flux2KleinPipeline

dtype = torch.bfloat16
device = "cuda"

height = 768
width = 1360
prompt = "a photo of a forest with mist swirling around the tree trunks. The word 'FLUX.2' is painted over it in big, r
ed brush strokes with visible texture"

repo_id = "black-forest-labs/FLUX.2-klein-9B"
pipeline = Flux2KleinPipeline.from_pretrained(repo_id, torch_dtype=dtype)
pipeline.to(device)
num_inference_steps = 4
guidance_scale = 1.0
generator = torch.Generator(device=device).manual_seed(42)

out = pipeline(
    prompt=prompt,
    height=height,
    width=width,
    num_inference_steps=num_inference_steps,
    guidance_scale=guidance_scale,
    generator=generator
).images[0]

out.save("output_9b.png")
```
output_9b

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by david6666666** [comment]
@ZJY0516 @SamitHuang @hsliuustc0106 PTAL thanks.

File: vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py [hunk 1]
```diff
@@ -0,0 +1,654 @@
...
+        value = value.unflatten(-1, (self.heads, -1))
+
+        query = self.norm_q(query)
+        key = self.norm_k(key)
+
+        if image_rotary_emb is not None:
+            query = apply_rotary_emb(query, image_rotary_emb, sequence_dim=1)
+            key = apply_rotary_emb(key, image_rotary_emb, sequence_dim=1)
+
+        attn_output = _scaled_dot_product_attention(query, key, value, attention_mask=attention_mask)
```
Comment (ZJY0516) [inline vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: please use attention layer in vllm_omni/diffusion

File: vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py [hunk 2]
```diff
@@ -0,0 +1,654 @@
...
+
+            encoder_query = self.norm_added_q(encoder_query)
+            encoder_key = self.norm_added_k(encoder_key)
+
+            query = torch.cat([encoder_query, query], dim=1)
+            key = torch.cat([encoder_key, key], dim=1)
+            value = torch.cat([encoder_value, value], dim=1)
+
+        if image_rotary_emb is not None:
+            query = apply_rotary_emb(query, image_rotary_emb, sequence_dim=1)
```
Comment (ZJY0516) [inline vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: we have rope layer in vllm-omni

File: vllm_omni/diffusion/request.py [hunk 3]
```diff
@@ -58,6 +58,7 @@ class OmniDiffusionRequest:
 
     # Additional text-related parameters
     max_sequence_length: int | None = None
+    text_encoder_out_layers: tuple[int, ...] | None = None
```
Comment (ZJY0516) [inline vllm_omni/diffusion/request.py]: why we need this?

File: vllm_omni/diffusion/cache/cache_dit_backend.py [hunk 4]
```diff
@@ -257,6 +257,75 @@ def refresh_cache_context(pipeline: Any, num_inference_steps: int, verbose: bool
     return refresh_cache_context
 
 
+def enable_cache_for_flux2_klein(pipeline: Any, cache_config: Any) -> Callable[[int], None]:
```
Comment (ZJY0516) [inline vllm_omni/diffusion/cache/cache_dit_backend.py]: if cache-dit has already supported it, do we still need this? @SamitHuang

File: vllm_omni/diffusion/models/flux2_klein/pipeline_flux2_klein.py [hunk 5]
```diff
@@ -0,0 +1,1036 @@
...
+        timesteps = scheduler.timesteps
+        num_inference_steps = len(timesteps)
+    else:
+        scheduler.set_timesteps(num_inference_steps, device=device, **kwargs)
+        timesteps = scheduler.timesteps
+    return timesteps, num_inference_steps
+
+
+# Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion_img2img.retrieve_latents
+def retrieve_latents(
```
Comment (SamitHuang) [inline vllm_omni/diffusion/models/flux2_klein/pipeline_flux2_klein.py]: why not just import

File: vllm_omni/diffusion/models/flux2_klein/pipeline_flux2_klein.py [hunk 6]
```diff
@@ -0,0 +1,1036 @@
...
+
+    a = (m_200 - m_10) / 190.0
+    b = m_200 - 200.0 * a
+    mu = a * num_steps + b
+
+    return float(mu)
+
+
+# Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.retrieve_timesteps
+def retrieve_timesteps(
```
Comment (SamitHuang) [inline vllm_omni/diffusion/models/flux2_klein/pipeline_flux2_klein.py]: ditto

File: vllm_omni/diffusion/cache/cache_dit_backend.py (hunk 4 above)
Comment (SamitHuang) [reply vllm_omni/diffusion/cache/cache_dit_backend.py]: yes because it's not a regular dit. but we need to test it. @david6666666

File: vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py [hunk 7]
```diff
@@ -0,0 +1,670 @@
...
+        super().__init__()
+        self.head_dim = dim_head
+        self.inner_dim = out_dim if out_dim is not None else dim_head * heads
+        self.query_dim = query_dim
+        self.out_dim = out_dim if out_dim is not None else query_dim
+        self.heads = out_dim // dim_head if out_dim is not None else heads
+        self.dropout = dropout
+        self.added_kv_proj_dim = added_kv_proj_dim
+
+        self.to_q = nn.Linear(query_dim, self.inner_dim, bias=bias)
```
Comment (SamitHuang) [inline vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: why not use `from vllm.model_executor.layers.linear import QKVParallelLinear, ReplicatedLinear`?

File: vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py [hunk 8]
```diff
@@ -0,0 +1,670 @@
...
+        self.to_k = nn.Linear(query_dim, self.inner_dim, bias=bias)
+        self.to_v = nn.Linear(query_dim, self.inner_dim, bias=bias)
+
+        self.norm_q = nn.RMSNorm(dim_head, eps=eps, elementwise_affine=elementwise_affine)
+        self.norm_k = nn.RMSNorm(dim_head, eps=eps, elementwise_affine=elementwise_affine)
+
+        self.to_out = nn.ModuleList([nn.Linear(self.inner_dim, self.out_dim, bias=out_bias), nn.Dropout(dropout)])
+
+        if added_kv_proj_dim is not None:
+            self.norm_added_q = nn.RMSNorm(dim_head, eps=eps)
```
Comment (SamitHuang) [inline vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: why not use RMSNorm from vllm?

Comment (ZJY0516) [reply vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: No need to use norm op in vllm here. native torch op + torch.compile will be better

**Comment by hsliuustc0106** [comment]
in terms of the seed, I remember @SamitHuang changed the default seed to 0 as it will influence some acc

**Comment by SamitHuang** [comment]
> in terms of the seed, I remember @SamitHuang changed the default seed to 0 as it will influence some acc

yes, but it doesn't influence this PR

File: vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py (hunk 8 above)
Comment (SamitHuang) [reply vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: vllm RMSNorm doesn't support torch.compile?

**Comment by hsliuustc0106** [comment]
any benchmark result for latency?

File: vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py (hunk 8 above)
Comment (ZJY0516) [reply vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: It does support. But the dispatch logic in vLLM is a little different with omni diffusion's custom op. So I think decounpling will better. And in most cases, vllm doesn't use custom kernel of RMSNorm

**Comment by david6666666** [comment]
> any benchmark result for latency?

H800 9B 0.8s, 4B 0.5s

**Comment by david6666666** [comment]
PR is ready

**Comment by david6666666** [comment]
[vllm-omni-amd-ci](https://buildkite.com/vllm/vllm-omni-amd-ci/builds/751) failed is not related

**Comment by jannikstdl** [comment]
Will is someday support Image editing in the API?

**Comment by ZJY0516** [comment]
> Will is someday support Image editing in the API?

I think we already support online seving for image edit, see https://github.com/vllm-project/vllm-omni/tree/main/examples/online_serving/image_to_image

**Comment by jannikstdl** [comment]
@ZJY0516 Uhm ok, will try in the latest [VLLM Omni Docker Image](http://hub.docker.com/r/vllm/vllm-omni/tags)

Could not find anything regarding to that in the docs.

File: vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py (hunk 8 above)
Comment (david6666666) [reply vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: I test native torch op + torch.compile is better

**Comment by kolmogorov-quyet** [comment]
They don't work with parallelism acceleration. Can anyone write a code that supports one of the four Ulysses-SP, Ring-SP, CFG-Parallel, or Tensor-Parallel architectures? I've tried one H2O GPU vs two H2O GPUs and the speed remains the same. Model FLUX.2-klein-9B, tested --usp 2, --ring 2(hybrid), Tensor-Parallel

**Comment by hsliuustc0106** [comment]
> They don't work with parallelism acceleration. Can anyone write a code that supports one of the four Ulysses-SP, Ring-SP, CFG-Parallel, or Tensor-Parallel architectures? I've tried one H2O GPU vs two H2O GPUs and the speed remains the same. Model FLUX.2-klein-9B, tested --usp 2, --ring 2(hybrid), Tensor-Parallel

@wtomin @ZJY0516 PTAL

**Comment by ZJY0516** [comment]
> They don't work with parallelism acceleration. Can anyone write a code that supports one of the four Ulysses-SP, Ring-SP, CFG-Parallel, or Tensor-Parallel architectures? I've tried one H2O GPU vs two H2O GPUs and the speed remains the same. Model FLUX.2-klein-9B, tested --usp 2, --ring 2(hybrid), Tensor-Parallel

Could you please try https://github.com/vllm-project/vllm-omni/pull/973?
